File: sdks/python/agentreel/costs.py
"""Model cost estimation (USD per 1M tokens, as of March 2026)."""

MODEL_COSTS: dict[str, dict[str, float]] = {
    # OpenAI
    "gpt-4.5-turbo": {"input": 5.00, "output": 15.00},
    "gpt-4o": {"input": 2.50, "output": 10.00},
    "gpt-4o-mini": {"input": 0.15, "output": 0.60},
    "o3": {"input": 10.00, "output": 40.00},
    "o3-mini": {"input": 1.10, "output": 4.40},
    "o4-mini": {"input": 1.10, "output": 4.40},
    # Anthropic
    "claude-opus-4-6": {"input": 15.00, "output": 75.00},
    "claude-sonnet-4-6": {"input": 3.00, "output": 15.00},
    "claude-sonnet-4-20250514": {"input": 3.00, "output": 15.00},
    "claude-haiku-4-5-20251001": {"input": 0.80, "output": 4.00},
    # Google
    "gemini-2.5-pro": {"input": 1.25, "output": 10.00},
    "gemini-2.5-flash": {"input": 0.15, "output": 0.60},
    # DeepSeek
    "deepseek-v3": {"input": 0.27, "output": 1.10},
    "deepseek-r1": {"input": 0.55, "output": 2.19},
}


def estimate_cost(model: str, input_tokens: int, output_tokens: int) -> float:
    """Estimate cost in USD for a given model and token count."""
    costs = MODEL_COSTS.get(model)
    if not costs:
        for key, val in MODEL_COSTS.items():
            if key in model or model in key:
                costs = val
                break
    if not costs:
        return 0.0
    return (input_tokens * costs["input"] / 1_000_000) + (
        output_tokens * costs["output"] / 1_000_000
    )

File: crates/agentreel-proxy/src/costs.rs
/// Model cost estimation (USD per 1M tokens, as of March 2026).

pub struct ModelCost {
    pub input: f64,
    pub output: f64,
}

/// Estimate cost in USD for a given model and token count.
pub fn estimate_cost(model: &str, input_tokens: u64, output_tokens: u64) -> f64 {
    let costs = match_model_cost(model);
    (input_tokens as f64 * costs.input / 1_000_000.0)
        + (output_tokens as f64 * costs.output / 1_000_000.0)
}

fn match_model_cost(model: &str) -> ModelCost {
    let m = model.to_lowercase();

    // Anthropic
    if m.contains("opus") {
        return ModelCost { input: 15.0, output: 75.0 };
    }
    if m.contains("sonnet") {
        return ModelCost { input: 3.0, output: 15.0 };
    }
    if m.contains("haiku") {
        return ModelCost { input: 0.80, output: 4.0 };
    }

    // OpenAI
    if m.contains("gpt-4.5") {
        return ModelCost { input: 5.0, output: 15.0 };
    }
    if m.contains("gpt-4o-mini") {
        return ModelCost { input: 0.15, output: 0.60 };
    }
    if m.contains("gpt-4o") {
        return ModelCost { input: 2.50, output: 10.0 };
    }
    if m.contains("o3-mini") || m.contains("o4-mini") {
        return ModelCost { input: 1.10, output: 4.40 };
    }
    if m.contains("o3") {
        return ModelCost { input: 10.0, output: 40.0 };
    }

    // Google
    if m.contains("gemini-2.5-pro") {
        return ModelCost { input: 1.25, output: 10.0 };
    }
    if m.contains("gemini") {
        return ModelCost { input: 0.15, output: 0.60 };
    }

    // DeepSeek
    if m.contains("deepseek-r1") {
        return ModelCost { input: 0.55, output: 2.19 };
    }
    if m.contains("deepseek") {
        return ModelCost { input: 0.27, output: 1.10 };
    }

    // Unknown — assume mid-range
    ModelCost { input: 3.0, output: 15.0 }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_claude_opus_cost() {
        let cost = estimate_cost("claude-opus-4-6", 1000, 500);
        assert!((cost - 0.0525).abs() < 0.0001);
    }

    #[test]
    fn test_claude_sonnet_cost() {
        let cost = estimate_cost("claude-sonnet-4-20250514", 1000, 500);
        assert!((cost - 0.0105).abs() < 0.0001);
    }

    #[test]
    fn test_gpt4o_cost() {
        let cost = estimate_cost("gpt-4o", 5000, 2000);
        assert!((cost - 0.0325).abs() < 0.0001);
    }

    #[test]
    fn test_unknown_model_uses_default() {
        let cost = estimate_cost("some-unknown-model", 1000, 1000);
        assert!(cost > 0.0);
    }
}
